File: src/main/services/GamebananaService.ts
import fs from 'fs'
import os from 'os'
import path from 'path'
import crypto from 'crypto'
import { net } from 'electron'
import type {
  GamebananaAlternateSourceDto,
  GamebananaCommentDto,
  GamebananaFileDto,
  GamebananaRequirementDto,
  GamebananaSearchResult,
  GamebananaSubmissionDto
} from '../../shared/types'
import { BALDI_COMMUNITY_CATEGORY_ID } from '../../shared/types'
import { debugLog } from '../logger'



const API_BASE = 'https://gamebanana.com/apiv12/'
const SITE_BASE = 'https://gamebanana.com'

const PAGE_SIZE = 50


const USER_AGENT = "GottaManageDev/0.1.0 (mod manager for Baldi's Basics Plus; electron)"



const BLOCKED_MOD_IDS = new Set<number>([675111])

type Json = Record<string, unknown>

function ok(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value)
}

function asNum(v: unknown, fallback = 0): number {
  const n = typeof v === 'number' ? v : fallback
  return Number.isFinite(n) ? n : fallback
}

function asStr(v: unknown, fallback = ''): string {
  return v === undefined || v === null ? fallback : String(v)
}


function absoluteUrl(u: string): string {
  if (!u) return u
  return u.startsWith('http') ? u : u.startsWith('/') ? SITE_BASE + u : u
}


function resolveLink(cloaked: string): string {
  if (!cloaked) return cloaked
  if (/\/\/gamebanana\.com\/linkfilter/i.test(cloaked)) {
    try {
      const q = new URL(cloaked).searchParams.get('url')
      if (q) return q
    } catch {
      
    }
  }
  return cloaked
}


function extractGamebananaId(u: string | undefined): number | undefined {
  if (!u) return undefined
  const real = resolveLink(u)
  const m = /\/mods\/0*(\d+)/i.exec(real)
  if (!m) return undefined
  const id = Number(m[1])
  return Number.isFinite(id) && id > 0 ? id : undefined
}


function hostOf(u: string): string {
  try {
    return new URL(u).hostname.replace(/^www\./, '')
  } catch {
    return ''
  }
}


function singleImageUrl(img: unknown): string | undefined {
  if (!ok(img)) return undefined
  const base = asStr(img['_sBaseUrl']).replace(/\/+$/, '')
  if (!base) return undefined
  const fileName =
    asStr(img['_sFile530']) || asStr(img['_sFile']) || asStr(img['_sFile100'])
  return fileName ? `${base}/${fileName}` : undefined
}

function parseThumb(media: unknown): string | undefined {
  if (!ok(media)) return undefined
  const images = media['_aImages']
  if (!Array.isArray(images) || images.length === 0) return undefined
  return singleImageUrl(images[0])
}


function parseAllImages(media: unknown): string[] {
  if (!ok(media)) return []
  const images = media['_aImages']
  if (!Array.isArray(images)) return []
  const out: string[] = []
  for (const img of images) {
    const u = singleImageUrl(img)
    if (u && !out.includes(u)) out.push(u)
  }
  return out
}


function parseFileVersion(desc: string): string | undefined {
  if (!desc) return undefined
  
  const m = /\(([^)]+)\)\s*$/.exec(desc) ?? /\(([^)]+)\)/.exec(desc)
  if (!m) return undefined
  const v = m[1].trim()
  return v || undefined
}

function parseFiles(node: unknown): GamebananaFileDto[] {
  const out: GamebananaFileDto[] = []
  const push = (arr: unknown): void => {
    if (!Array.isArray(arr)) return
    for (const it of arr) {
      if (!ok(it)) continue
      const desc = asStr(it['_sDescription'])
      const fileSize = asNum(it['_nFilesize'])
      out.push({
        id: asNum(it['_idRow']),
        fileName: asStr(it['_sFile']),
        fileSize,
        downloadUrl: absoluteUrl(asStr(it['_sDownloadUrl'])),
        description: desc || undefined,
        version: parseFileVersion(desc),
        dateAdded: asNum(it['_tsDateAdded']) || undefined
      })
    }
  }
  push(node)
  return out.filter((f) => f.downloadUrl)
}





async function request(
  url: string,
  init: { headers?: Record<string, string>; signal?: AbortSignal }
): Promise<Response> {
  
  for (let attempt = 0; ; attempt++) {
    try {
      let res: Response | undefined
      if (net.isOnline?.()) {
        try {
          res = await net.fetch(url, {
            headers: init.headers,
            redirect: 'follow',
            signal: init.signal
          })
        } catch {
          
        }
      }
      if (!res) res = await fetch(url, { headers: init.headers, redirect: 'follow', signal: init.signal })
      return res
    } catch (err) {
      const aborted = init.signal?.aborted || (err instanceof DOMException && err.name === 'AbortError')
      if (aborted || attempt >= 2) throw err
      await new Promise((resolve) => setTimeout(resolve, 400 * (attempt + 1)))
    }
  }
}

async function getJson(url: string): Promise<Json> {
  const res = await request(url, {
    headers: { 'User-Agent': USER_AGENT, Accept: 'application/json' }
  })
  if (!res.ok) {
    throw new Error(`GameBanana API ${res.status}: ${res.statusText}`)
  }
  return (await res.json()) as Json
}






export async function searchMods(
  page: number,
  query?: string,
  category?: number
): Promise<GamebananaSearchResult> {
  const q = query ? query.trim() : ''
  const categoryId = category && category > 0 ? category : BALDI_COMMUNITY_CATEGORY_ID
  const filters: string[] = [`_aFilters[Generic_Category]=${categoryId}`]
  if (q) filters.push(`_aFilters[Generic_Name]=contains,${encodeURIComponent(q)}`)
  const url =
    `${API_BASE}Mod/Index?_nPerpage=${PAGE_SIZE}` +
    `&${filters.join('&')}&_nPage=${Math.max(1, page)}`
  debugLog('searchMods url =', url)

  const doc = await getJson(url)
  const records = doc['_aRecords']
  const items: GamebananaSubmissionDto[] = []
  if (Array.isArray(records)) {
    for (const r of records) {
      if (!ok(r)) continue
      const id = asNum(r['_idRow'])
      
      if (BLOCKED_MOD_IDS.has(id)) continue
      const preview = r['_aPreviewMedia']
      const submitter = ok(r['_aSubmitter']) ? r['_aSubmitter'] : undefined
      const category = ok(r['_aCategory']) ? r['_aCategory'] : undefined
      items.push({
        id,
        name: asStr(r['_sName']),
        version: asStr(r['_sVersion']) || undefined,
        authorName: ok(submitter) ? asStr(submitter['_sName']) || undefined : undefined,
        hasFiles: r['_bHasFiles'] === true,
        categoryId: category ? asNum(category['_idRow']) || undefined : undefined,
        thumbnailUrl: parseThumb(preview),
        
        viewCount: asNum(r['_nViewCount']) || undefined,
        dateAdded: asNum(r['_tsDateAdded']) || undefined,
        dateUpdated: asNum(r['_tsDateModified']) || asNum(r['_tsDateUpdated']) || undefined,
        files: []
      })
    }
  }

  
  await Promise.allSettled(
    items.map(async (it) => {
      try {
        const doc = await getJson(API_BASE + 'Mod/' + it.id + '/ProfilePage')
        it.downloadCount = asNum(doc['_nDownloadCount']) || undefined
      } catch {
        
      }
    })
  )

  const meta = ok(doc['_aMetadata']) ? doc['_aMetadata'] : undefined
  const recordCount = meta ? asNum(meta['_nRecordCount']) : items.length
  return {
    recordCount,
    isComplete: meta ? meta['_bIsComplete'] === true : items.length < PAGE_SIZE,
    perPage: meta ? asNum(meta['_nPerpage'], PAGE_SIZE) : PAGE_SIZE,
    items
  }
}






function parseRequirements(raw: unknown): GamebananaRequirementDto[] {
  if (!Array.isArray(raw)) return []
  const out: GamebananaRequirementDto[] = []
  for (const item of raw) {
    
    const entry = Array.isArray(item)
      ? item
      : ok(item) && Array.isArray(item['value'])
        ? (item['value'] as unknown[])
        : null
    if (!entry || entry.length === 0) continue
    const name = asStr(entry[0])
    if (!name) continue
    
    const origUrl = asStr(entry[1]) || undefined
    const url = origUrl ? absoluteUrl(resolveLink(origUrl)) : undefined
    const status = asStr(entry[2]) || undefined
    const reqLevel = asStr(entry[4]) || ''
    out.push({
      name,
      url,
      status,
      required: /required/i.test(reqLevel),
      gamebananaId: extractGamebananaId(origUrl)
    })
  }
  return out
}


function parseAlternateFileSources(raw: unknown): GamebananaAlternateSourceDto[] {
  if (!Array.isArray(raw)) return []
  const out: GamebananaAlternateSourceDto[] = []
  for (const src of raw) {
    if (!ok(src)) continue
    const origUrl = asStr(src['url'])
    if (!origUrl) continue
    const real = resolveLink(origUrl)
    out.push({
      url: absoluteUrl(real),
      description: asStr(src['description']) || undefined,
      host: hostOf(real)
    })
  }
  return out
}





export async function getSubmission(
  id: number
): Promise<GamebananaSubmissionDto> {
  if (!Number.isFinite(id) || id <= 0) throw new Error('Invalid submission id')
  const doc = await getJson(`${API_BASE}Mod/${id}/ProfilePage`)
  debugLog('getSubmission id =', id, 'name =', asStr(doc['_sName']), 'categoryId =', ok(doc['_aCategory']) ? asNum(doc['_aCategory']['_idRow']) : 'n/a')

  
  const files = parseFiles(doc['_aFiles'])
  const archivedFiles = parseFiles(doc['_aArchivedFiles'])
  const submitter = ok(doc['_aSubmitter']) ? doc['_aSubmitter'] : undefined
  
  const primaryDl = absoluteUrl(asStr(doc['_sDownloadUrl']))
  if (primaryDl && !files.some((f) => f.downloadUrl === primaryDl)) {
    files.unshift({ id: -1, fileName: asStr(doc['_sName']) + '.zip', fileSize: 0, downloadUrl: primaryDl })
  }

  const images = parseAllImages(doc['_aPreviewMedia'])
  const category = ok(doc['_aCategory']) ? doc['_aCategory'] : undefined
  return {
    id: asNum(doc['_idRow'], id),
    name: asStr(doc['_sName']),
    
    description: asStr(doc['_sText']) || asStr(doc['_sDescription']) || undefined,
    version: asStr(doc['_sVersion']) || undefined,
    authorName: ok(submitter) ? asStr(submitter['_sName']) || undefined : undefined,
    hasFiles: files.length > 0 || archivedFiles.length > 0,
    categoryId: category ? asNum(category['_idRow']) || undefined : undefined,
    thumbnailUrl: images[0] ?? parseThumb(doc['_aPreviewMedia']),
    images,
    downloadCount: asNum(doc['_nDownloadCount']) || undefined,
    viewCount: asNum(doc['_nViewCount']) || undefined,
    dateAdded: asNum(doc['_tsDateAdded']) || undefined,
    dateUpdated: asNum(doc['_tsDateModified']) || asNum(doc['_tsDateUpdated']) || undefined,
    files,
    archivedFiles,
    requirements: parseRequirements(doc['_aRequirements']),
    alternateFileSources: parseAlternateFileSources(doc['_aAlternateFileSources'])
  }
}





export async function getComments(submissionId: number): Promise<GamebananaCommentDto[]> {
  if (!Number.isFinite(submissionId) || submissionId <= 0) return []
  try {
    
    const doc = await getJson(`${API_BASE}Mod/${submissionId}/Posts`)
    const records = doc['_aRecords']
    if (!Array.isArray(records)) return []
    const out: GamebananaCommentDto[] = []
    for (const r of records) {
      const c = parseComment(r)
      if (c) out.push(c)
    }
    return out
  } catch {
    return []
  }
}





export async function getPostReplies(postId: number): Promise<GamebananaCommentDto[]> {
  if (!Number.isFinite(postId) || postId <= 0) return []
  try {
    const doc = await getJson(`${API_BASE}Post/${postId}/Posts?_nPage=1&_nPerpage=20`)
    const records = doc['_aRecords']
    if (!Array.isArray(records)) return []
    const out: GamebananaCommentDto[] = []
    for (const r of records) {
      const c = parseComment(r)
      if (c) out.push(c)
    }
    return out.reverse()
  } catch {
    return []
  }
}


function parseComment(r: unknown): GamebananaCommentDto | null {
  if (!ok(r)) return null
  const poster = ok(r['_aPoster']) ? r['_aPoster'] : undefined
  const ts = asNum(r['_tsDateAdded'])
  return {
    id: asNum(r['_idRow']),
    author: ok(poster) ? asStr(poster['_sName']) || 'Unknown' : 'Unknown',
    body: asStr(r['_sText']),
    date: ts ? new Date(ts * 1000).toLocaleString() : undefined,
    replyCount: asNum(r['_nReplyCount'])
  }
}

export interface DownloadProgress {
  received: number
  total?: number
}





export async function downloadMod(
  downloadUrl: string,
  onProgress?: (p: DownloadProgress) => void,
  isCancelled?: () => boolean
): Promise<string> {
  const res = await request(downloadUrl, {
    headers: { 'User-Agent': USER_AGENT }
  })
  if (!res.ok || !res.body) {
    throw new Error(`Download failed ${res.status}: ${res.statusText}`)
  }

  const total = res.headers.get('content-length')
  const totalBytes = total ? Number(total) : undefined
  const tmpDir = fs.mkdtempSync(path.join(os.tmpdir(), 'gmdl-'))
  const name =
    (downloadUrl.split('/').pop() || 'mod') +
    '-' +
    crypto.randomBytes(4).toString('hex')
  const dest = path.join(tmpDir, name)

  const stream = fs.createWriteStream(dest)
  const reader = res.body.getReader()
  let received = 0
  try {
    for (;;) {
      if (isCancelled?.()) {
        throw new Error('Download cancelled')
      }
      const { done, value } = await reader.read()
      if (done) break
      if (value) {
        received += value.byteLength
        if (!stream.write(value)) {
          await new Promise<void>((resolve) => stream.once('drain', () => resolve(undefined)))
        }
        onProgress?.({ received, total: totalBytes })
      }
    }
  } finally {
    reader.releaseLock?.()
    stream.end()
    await new Promise<void>((resolve) => stream.once('finish', () => resolve(undefined)))
  }
  return dest
}

File: RBBPMM.Wpf/RBBPMM.Core/GamebananaService.cs
using System.Text.Json;
using System.Text.Json.Nodes;
using System.Text.RegularExpressions;

namespace RBBPMM.Core;

/// <summary>
/// GameBanana apiv12 client. Ported from src/main/services/GamebananaService.ts.
/// The HttpClient is injectable so the parsing/path logic can be unit-tested without the network.
/// </summary>
public sealed class GamebananaService
{
    public const string ApiBase = "https://gamebanana.com/apiv12/";
    public const string SiteBase = "https://gamebanana.com";
    public const int PageSize = 50;
    public const string UserAgent = "GottaManageDev/0.1.0 (mod manager for Baldi's Basics Plus; dotnet)";

    private static readonly HashSet<int> BlockedModIds = [675111];

    private static readonly Regex[] IdPatterns =
    [
        new(@"/(?:mods|wip|m)/0*(\d+)", RegexOptions.IgnoreCase),
        new(@"/details/(?:[^/]+\/)*0*(\d+)", RegexOptions.IgnoreCase),
        new(@"[?&](?:id|modsid|submissionid)=(\d+)", RegexOptions.IgnoreCase),
    ];

    private readonly HttpClient _http;

    public GamebananaService(HttpClient? httpClient = null)
    {
        _http = httpClient ?? new HttpClient { Timeout = TimeSpan.FromSeconds(20) };
        if (_http.DefaultRequestHeaders.UserAgent.Count == 0)
            _http.DefaultRequestHeaders.UserAgent.ParseAdd(UserAgent);
    }

    // ---- low-level req ----

    public async Task<JsonNode> GetJsonAsync(string url, CancellationToken ct = default)
    {
        using var resp = await _http.GetAsync(url, ct);
        if (!resp.IsSuccessStatusCode)
            throw new HttpRequestException($"GameBanana API {(int)resp.StatusCode}: {resp.ReasonPhrase}");
        var body = await resp.Content.ReadAsStringAsync(ct);
        return JsonNode.Parse(body) ?? throw new InvalidOperationException("Empty GameBanana response");
    }

    public async Task<string> DownloadFileAsync(string downloadUrl, string destDir, string? fileName = null,
        IProgress<(long Received, long? Total)>? progress = null, Func<bool>? isCancelled = null)
    {
        using var resp = await _http.GetAsync(downloadUrl, HttpCompletionOption.ResponseHeadersRead);
        if (!resp.IsSuccessStatusCode || resp.Content == null)
            throw new HttpRequestException($"Download failed {(int)resp.StatusCode}: {resp.ReasonPhrase}");
        var total = resp.Content.Headers.ContentLength;
        var cd = resp.Content.Headers.ContentDisposition?.FileName
                 ?? resp.Content.Headers.ContentDisposition?.FileNameStar;
        var baseName = fileName
            ?? ExtractFilenameFromCd(resp.Content.Headers.ContentDisposition?.ToString())
            ?? Path.GetFileName(new Uri(downloadUrl).LocalPath);
        if (string.IsNullOrEmpty(baseName)) baseName = "mod";
        baseName = SanitizeFileName(baseName);
        Directory.CreateDirectory(destDir);
        var dest = Path.Combine(destDir, baseName);
        await using var outStream = File.Create(dest);
        await using var inStream = await resp.Content.ReadAsStreamAsync();
        var buffer = new byte[8192];
        long received = 0;
        int read;
        while ((read = await inStream.ReadAsync(buffer, 0, buffer.Length, default)) > 0)
        {
            if (isCancelled?.Invoke() == true) throw new OperationCanceledException("Download cancelled");
            await outStream.WriteAsync(buffer.AsMemory(0, read));
            received += read;
            progress?.Report((received, total));
        }
        return dest;
    }

    // ---- endpoints ----

    public async Task<GamebananaSearchResult> SearchMods(int page, string? query = null, int? category = null,
        CancellationToken ct = default)
    {
        var q = query?.Trim() ?? "";
        var categoryId = category is > 0 ? category.Value : CategoryIds.BaldiCommunity;
        var filters = new List<string> { $"_aFilters[Generic_Category]={categoryId}" };
        if (!string.IsNullOrEmpty(q))
            filters.Add($"_aFilters[Generic_Name]=contains,{Uri.EscapeDataString(q)}");
        var url = $"{ApiBase}Mod/Index?_nPerpage={PageSize}&{string.Join("&", filters)}&_nPage={Math.Max(1, page)}";
        var doc = await GetJsonAsync(url, ct);
        var records = doc["_aRecords"];
        var items = new List<GamebananaSubmission>();
        if (records is JsonArray arr)
        {
            foreach (var r in arr)
            {
                if (r is not JsonObject o) continue;
                var id = AsNum(o["_idRow"]);
                if (BlockedModIds.Contains((int)id)) continue;
                var preview = o["_aPreviewMedia"];
                var submitter = o["_aSubmitter"] as JsonObject;
                var categoryNode = o["_aCategory"] as JsonObject;
                items.Add(new GamebananaSubmission
                {
                    Id = (int)id,
                    Name = AsStr(o["_sName"]),
                    Version = AsStr(o["_sVersion"]),
                    AuthorName = submitter?["_sName"]?.GetValue<string>(),
                    HasFiles = o["_bHasFiles"]?.GetValue<bool>() ?? false,
                    CategoryId = categoryNode?["_idRow"]?.GetValue<int>(),
                    ThumbnailUrl = ParseThumb(preview),
                    ViewCount = AsNumNullable(o["_nViewCount"]),
                    DateAdded = AsNumNullable(o["_tsDateAdded"]),
                    DateUpdated = AsNumNullable(o["_tsDateModified"]) ?? AsNumNullable(o["_tsDateUpdated"]),
                    Files = []
                });
            }
        }
        await Task.WhenAll(items.Select(async it =>
        {
            try
            {
                var d = await GetJsonAsync($"{ApiBase}Mod/{it.Id}/ProfilePage", ct);
                it.DownloadCount = AsNumNullable(d["_nDownloadCount"]);
            }
            catch { /* best effort */ }
        }));
        var meta = doc["_aMetadata"] as JsonObject;
        var recordCount = meta?["_nRecordCount"]?.GetValue<long>() ?? items.Count;
        var isComplete = meta?["_bIsComplete"]?.GetValue<bool>() ?? (items.Count < PageSize);
        var perPage = meta?["_nPerpage"]?.GetValue<int>() ?? PageSize;
        return new GamebananaSearchResult(recordCount, isComplete, perPage, items);
    }

    public async Task<GamebananaSubmission> GetSubmission(int id, CancellationToken ct = default)
    {
        if (id <= 0) throw new ArgumentException("Invalid submission id", nameof(id));
        var doc = await GetJsonAsync($"{ApiBase}Mod/{id}/ProfilePage", ct);
        var files = ParseFiles(doc["_aFiles"]);
        var archivedFiles = ParseFiles(doc["_aArchivedFiles"]);
        var submitter = doc["_aSubmitter"] as JsonObject;
        var primaryDl = AbsoluteUrl(AsStr(doc["_sDownloadUrl"]));
        if (!string.IsNullOrEmpty(primaryDl) && !files.Any(f => f.DownloadUrl == primaryDl))
            files.Insert(0, new GamebananaFile(-1, AsStr(doc["_sName"]) + ".zip", 0, primaryDl));
        var images = ParseAllImages(doc["_aPreviewMedia"]);
        var category = doc["_aCategory"] as JsonObject;
        return new GamebananaSubmission
        {
            Id = doc["_idRow"]?.GetValue<int>() ?? id,
            Name = AsStr(doc["_sName"]),
            Description = AsStr(doc["_sText"]) ?? AsStr(doc["_sDescription"]),
            Version = AsStr(doc["_sVersion"]),
            AuthorName = submitter?["_sName"]?.GetValue<string>(),
            HasFiles = files.Count > 0 || archivedFiles.Count > 0,
            CategoryId = category?["_idRow"]?.GetValue<int>(),
            ThumbnailUrl = images.Count > 0 ? images[0] : ParseThumb(doc["_aPreviewMedia"]),
            Images = images,
            DownloadCount = AsNumNullable(doc["_nDownloadCount"]),
            ViewCount = AsNumNullable(doc["_nViewCount"]),
            DateAdded = AsNumNullable(doc["_tsDateAdded"]),
            DateUpdated = AsNumNullable(doc["_tsDateModified"]) ?? AsNumNullable(doc["_tsDateUpdated"]),
            Files = files,
            ArchivedFiles = archivedFiles,
            Requirements = ParseRequirements(doc["_aRequirements"]),
            AlternateFileSources = ParseAlternateFileSources(doc["_aAlternateFileSources"])
        };
    }

    public async Task<GamebananaUpdates> GetUpdates(int submissionId, CancellationToken ct = default)
    {
        if (submissionId <= 0) return new GamebananaUpdates(0, []);
        try
        {
            var doc = await GetJsonAsync($"{ApiBase}Mod/{submissionId}/Updates?_nPerpage={PageSize}&_nPage=1", ct);
            var records = doc["_aRecords"];
            var meta = doc["_aMetadata"] as JsonObject;
            var items = new List<GamebananaUpdate>();
            if (records is JsonArray arr)
                foreach (var r in arr)
                {
                    var u = ParseUpdate(r);
                    if (u != null) items.Add(u);
                }
            var total = meta?["_nRecordCount"]?.GetValue<long>() ?? items.Count;
            return new GamebananaUpdates(total, items);
        }
        catch { return new GamebananaUpdates(0, []); }
    }

    public async Task<GamebananaComments> GetComments(int submissionId, CancellationToken ct = default)
    {
        if (submissionId <= 0) return new GamebananaComments(0, []);
        try
        {
            var doc = await GetJsonAsync($"{ApiBase}Mod/{submissionId}/Posts", ct);
            var records = doc["_aRecords"];
            var meta = doc["_aMetadata"] as JsonObject;
            var total = meta?["_nRecordCount"]?.GetValue<long>() ?? 0;
            var items = new List<GamebananaComment>();
            if (records is JsonArray arr)
                foreach (var r in arr)
                {
                    var c = ParseComment(r);
                    if (c != null) items.Add(c);
                }
            return new GamebananaComments(total, items);
        }
        catch { return new GamebananaComments(0, []); }
    }

    public async Task<List<GamebananaComment>> GetPostReplies(int postId, CancellationToken ct = default)
    {
        if (postId <= 0) return [];
        try
        {
            var doc = await GetJsonAsync($"{ApiBase}Post/{postId}/Posts?_nPage=1&_nPerpage=20", ct);
            var records = doc["_aRecords"];
            if (records is not JsonArray arr) return [];
            var outList = new List<GamebananaComment>();
            foreach (var r in arr)
            {
                var c = ParseComment(r);
                if (c != null) outList.Add(c);
            }
            outList.Reverse();
            return outList;
        }
        catch { return []; }
    }

    // ---- pure parse helpers (exposed for unit tests) ----

    public static int? ExtractGamebananaId(string? u)
    {
        if (string.IsNullOrEmpty(u)) return null;
        var real = ResolveLink(u);
        var baseUrl = real.Split('?', '#')[0];
        foreach (var p in IdPatterns)
        {
            var hay = p.ToString().Contains('?') ? real : baseUrl;
            var m = p.Match(hay);
            if (!m.Success) continue;
            if (int.TryParse(m.Groups[m.Groups.Count - 1].Value, out var id) && id > 0)
                return id;
        }
        return null;
    }

    public static string AbsoluteUrl(string u)
    {
        if (string.IsNullOrEmpty(u)) return u;
        return u.StartsWith("http") ? u : u.StartsWith('/') ? SiteBase + u : u;
    }

    public static string ResolveLink(string cloaked)
    {
        if (string.IsNullOrEmpty(cloaked)) return cloaked;
        if (Regex.IsMatch(cloaked, @"//gamebanana\.com/linkfilter", RegexOptions.IgnoreCase))
        {
            try
            {
                var q = new Uri(cloaked).Query;
                var idx = q.IndexOf("url=", StringComparison.OrdinalIgnoreCase);
                if (idx >= 0)
                {
                    var v = q[(idx + 4)..];
                    var amp = v.IndexOf('&');
                    if (amp >= 0) v = v[..amp];
                    var decoded = Uri.UnescapeDataString(v);
                    if (!string.IsNullOrEmpty(decoded)) return decoded;
                }
            }
            catch { /* ignore */ }
        }
        return cloaked;
    }

    public static List<GamebananaFile> ParseFiles(JsonNode? node)
    {
        var outList = new List<GamebananaFile>();
        if (node is not JsonArray arr) return outList;
        foreach (var it in arr)
        {
            if (it is not JsonObject o) continue;
            var desc = AsStr(o["_sDescription"]);
            outList.Add(new GamebananaFile(
                (int)AsNum(o["_idRow"]),
                AsStr(o["_sFile"]),
                AsNum(o["_nFilesize"]),
                AbsoluteUrl(AsStr(o["_sDownloadUrl"])),
                string.IsNullOrEmpty(desc) ? null : desc,
                ParseFileVersion(desc),
                AsNumNullable(o["_tsDateAdded"])));
        }
        return outList.Where(f => !string.IsNullOrEmpty(f.DownloadUrl)).ToList();
    }

    public static GamebananaUpdate? ParseUpdate(JsonNode? node)
    {
        if (node is not JsonObject r) return null;
        if (r["_bIsTrashed"]?.GetValue<bool>() == true || r["_bIsPrivate"]?.GetValue<bool>() == true)
            return null;
        var changeLog = new List<GamebananaUpdateChange>();
        if (r["_aChangeLog"] is JsonArray log)
            foreach (var e in log)
            {
                if (e is not JsonObject o) continue;
                var changeText = AsStr(o["text"]).Trim();
                if (string.IsNullOrEmpty(changeText)) continue;
                changeLog.Add(new GamebananaUpdateChange(changeText, AsStr(o["cat"]).Trim()));
            }
        var fileNames = new List<string>();
        if (r["_aFiles"] is JsonArray files)
            foreach (var f in files)
            {
                if (f is not JsonObject o) continue;
                var name = AsStr(o["_sFile"]).Trim();
                if (!string.IsNullOrEmpty(name) && !fileNames.Contains(name)) fileNames.Add(name);
            }
        var submitter = r["_aSubmitter"] as JsonObject;
        var text = AsStr(r["_sText"]).Trim();
        return new GamebananaUpdate(
            (int)AsNum(r["_idRow"]),
            AsStr(r["_sName"]).Trim(),
            AbsoluteUrl(AsStr(r["_sProfileUrl"])),
            AsNumNullable(r["_tsDateAdded"]),
            AsStr(r["_sVersion"]).Trim(),
            string.IsNullOrEmpty(text) ? null : text,
            submitter?["_sName"]?.GetValue<string>()?.Trim(),
            changeLog,
            fileNames);
    }

    public static GamebananaComment? ParseComment(JsonNode? node)
    {
        if (node is not JsonObject r) return null;
        var poster = r["_aPoster"] as JsonObject;
        var ts = AsNumNullable(r["_tsDateAdded"]);
        return new GamebananaComment(
            (int)AsNum(r["_idRow"]),
            poster?["_sName"]?.GetValue<string>() ?? "Unknown",
            AsStr(r["_sText"]),
            ts.HasValue ? DateTimeOffset.FromUnixTimeSeconds(ts.Value).ToLocalTime().ToString() : null,
            (int?)AsNumNullable(r["_nReplyCount"]));
    }

    public static string? ParseFileVersion(string desc)
    {
        if (string.IsNullOrEmpty(desc)) return null;
        var m = Regex.Match(desc, @"\(([^)]+)\)\s*$") != Match.Empty
            ? Regex.Match(desc, @"\(([^)]+)\)\s*$")
            : Regex.Match(desc, @"\(([^)]+)\)");
        if (!m.Success) return null;
        var v = m.Groups[1].Value.Trim();
        return string.IsNullOrEmpty(v) ? null : v;
    }

    public static string? ExtractFilenameFromCd(string? cd)
    {
        if (string.IsNullOrEmpty(cd)) return null;
        var utf8 = Regex.Match(cd, @"filename\*=UTF-8''([^;]+)", RegexOptions.IgnoreCase);
        if (utf8.Success)
        {
            try { return Uri.UnescapeDataString(utf8.Groups[1].Value.Trim().Trim('"')); }
            catch { /* ignore */ }
        }
        var plain = Regex.Match(cd, @"filename=""([^""]+)""", RegexOptions.IgnoreCase);
        if (!plain.Success) plain = Regex.Match(cd, @"filename=([^;]+)", RegexOptions.IgnoreCase);
        return plain.Success ? plain.Groups[1].Value.Trim().Trim('"') : null;
    }

    private static string? ParseThumb(JsonNode? media)
    {
        var imgs = ParseAllImages(media);
        return imgs.Count > 0 ? imgs[0] : null;
    }

    private static List<string> ParseAllImages(JsonNode? media)
    {
        var outList = new List<string>();
        if (media is not JsonObject o) return outList;
        if (o["_aImages"] is not JsonArray imgs) return outList;
        foreach (var img in imgs)
        {
            var u = SingleImageUrl(img);
            if (u != null && !outList.Contains(u)) outList.Add(u);
        }
        return outList;
    }

    private static string? SingleImageUrl(JsonNode? img)
    {
        if (img is not JsonObject o) return null;
        var baseUrl = AsStr(o["_sBaseUrl"]).TrimEnd('/');
        if (string.IsNullOrEmpty(baseUrl)) return null;
        var fileName = AsStr(o["_sFile530"]) ?? AsStr(o["_sFile"]) ?? AsStr(o["_sFile100"]);
        return string.IsNullOrEmpty(fileName) ? null : $"{baseUrl}/{fileName}";
    }

    private static List<GamebananaRequirement> ParseRequirements(JsonNode? raw)
    {
        var outList = new List<GamebananaRequirement>();
        if (raw is not JsonArray arr) return outList;
        foreach (var item in arr)
        {
            JsonArray? entry = item as JsonArray;
            if (item is JsonObject o && o["value"] is JsonArray v) entry = v;
            if (entry == null || entry.Count == 0) continue;
            var name = AsStr(entry[0]);
            if (string.IsNullOrEmpty(name)) continue;
            var origUrl = entry.Count > 1 ? AsStr(entry[1]) : null;
            var url = string.IsNullOrEmpty(origUrl) ? null : AbsoluteUrl(ResolveLink(origUrl));
            var status = entry.Count > 2 ? AsStr(entry[2]) : null;
            var reqLevel = entry.Count > 4 ? AsStr(entry[4]) : "";
            outList.Add(new GamebananaRequirement(name, url, status, Regex.IsMatch(reqLevel ?? "", "required", RegexOptions.IgnoreCase), ExtractGamebananaId(origUrl)));
        }
        return outList;
    }

    private static List<GamebananaAlternateSource> ParseAlternateFileSources(JsonNode? raw)
    {
        var outList = new List<GamebananaAlternateSource>();
        if (raw is not JsonArray arr) return outList;
        foreach (var src in arr)
        {
            if (src is not JsonObject o) continue;
            var origUrl = AsStr(o["url"]);
            if (string.IsNullOrEmpty(origUrl)) continue;
            var real = ResolveLink(origUrl);
            outList.Add(new GamebananaAlternateSource(AbsoluteUrl(real), AsStr(o["description"]), HostOf(real)));
        }
        return outList;
    }

    private static string? HostOf(string u)
    {
        try { return new Uri(u).Host.Replace("www.", ""); }
        catch { return ""; }
    }

    // ---- scalar coercion ----

    public static string AsStr(JsonNode? node, string fallback = "")
        => node == null || node.GetValueKind() == JsonValueKind.Null ? fallback : node.ToString() ?? fallback;

    public static long AsNum(JsonNode? node, long fallback = 0)
    {
        if (node == null) return fallback;
        if (node is JsonValue v)
        {
            if (v.TryGetValue<long>(out var l)) return l;
            if (v.TryGetValue<double>(out var d) && double.IsFinite(d)) return (long)d;
        }
        return fallback;
    }

    public static long? AsNumNullable(JsonNode? node)
    {
        if (node == null || node.GetValueKind() == JsonValueKind.Null) return null;
        if (node is JsonValue v)
        {
            if (v.TryGetValue<long>(out var l)) return l;
            if (v.TryGetValue<double>(out var d) && double.IsFinite(d)) return (long)d;
        }
        return null;
    }

    private static string SanitizeFileName(string name)
    {
        var invalid = Path.GetInvalidFileNameChars();
        var sb = new System.Text.StringBuilder(name.Length);
        foreach (var c in name)
            sb.Append(invalid.Contains(c) ? '_' : c);
        return sb.ToString();
    }
}
